File: python/price_axis.py
"""Deterministic y-axis ticks for price charts.

Reference implementation of SPEC.md. Standard library only.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Sequence

SPEC_VERSION = "1.0.0"

MANTISSAS = (1.0, 2.0, 5.0)
DEFAULT_MAX_TICKS = 8
DEFAULT_PADDING = 0.05
DEFAULT_MAX_DECIMALS = 6
FLAT_RATIO = 0.001
ZERO_SPAN = 1.0

_EPS = 1e-12
_LOG_NUDGE = 1e-9


class AxisError(ValueError):
    """The inputs cannot produce a valid axis. Reject, never repair."""


@dataclass(frozen=True)
class Axis:
    min: float
    max: float
    step: float
    decimals: int
    ticks: tuple
    labels: tuple
    flat: bool

    @property
    def count(self) -> int:
        return len(self.ticks)

    def as_dict(self) -> dict:
        return {
            "min": self.min,
            "max": self.max,
            "step": self.step,
            "decimals": self.decimals,
            "ticks": list(self.ticks),
            "labels": list(self.labels),
            "flat": self.flat,
        }


def _pow10(exp: int) -> float:
    """10**exp via the decimal literal.

    `_pow10(exp)` is a library pow and is not guaranteed to be the correctly rounded
    double for every exponent, so two languages can disagree in the last bit. Parsing
    the literal is correctly rounded everywhere, which keeps the implementations
    bit-identical. The JavaScript port uses Number("1e" + exp) for the same reason.
    """
    return float(f"1e{exp}")


def _exponent(value: float) -> int:
    """Base-10 exponent of a positive nice step, robust to log10 fuzz."""
    return math.floor(math.log10(value) + _LOG_NUDGE)


def nice_step(raw: float) -> float:
    """Smallest step of the form m*10^n, m in {1,2,5}, that is >= raw. R3."""
    if not math.isfinite(raw) or raw <= 0.0:
        raise AxisError(f"raw step must be finite and positive, got {raw!r}")
    exp = math.floor(math.log10(raw))
    for mantissa in MANTISSAS:
        candidate = mantissa * _pow10(exp)
        if candidate >= raw * (1.0 - _EPS):
            return candidate
    return _pow10(exp + 1)


def next_step(step: float) -> float:
    """The next nice step up: 1 -> 2 -> 5 -> 10. R5."""
    if not math.isfinite(step) or step <= 0.0:
        raise AxisError(f"step must be finite and positive, got {step!r}")
    exp = _exponent(step)
    mantissa = step / _pow10(exp)
    if mantissa < 1.5:
        return 2.0 * _pow10(exp)
    if mantissa < 3.5:
        return 5.0 * _pow10(exp)
    return _pow10(exp + 1)


def _check_number(name: str, value) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise AxisError(f"{name} must be a number, got {value!r}")
    value = float(value)
    if not math.isfinite(value):
        raise AxisError(f"{name} must be finite, got {value!r}")
    return value


def compute_axis(
    data_min,
    data_max,
    *,
    max_ticks: int = DEFAULT_MAX_TICKS,
    padding: float = DEFAULT_PADDING,
    min_decimals: int = 0,
    max_decimals: int = DEFAULT_MAX_DECIMALS,
) -> Axis:
    """Compute the axis for a price series. See SPEC.md."""
    data_min = _check_number("data_min", data_min)
    data_max = _check_number("data_max", data_max)
    padding = _check_number("padding", padding)

    if data_max < data_min:
        raise AxisError(f"data_max {data_max!r} is below data_min {data_min!r}")
    if not isinstance(max_ticks, int) or isinstance(max_ticks, bool) or max_ticks < 3:
        raise AxisError(f"max_ticks must be an integer >= 3, got {max_ticks!r}")
    if padding < 0.0:
        raise AxisError(f"padding must not be negative, got {padding!r}")
    for name, value in (("min_decimals", min_decimals), ("max_decimals", max_decimals)):
        if not isinstance(value, int) or isinstance(value, bool) or value < 0:
            raise AxisError(f"{name} must be a non-negative integer, got {value!r}")
    if min_decimals > max_decimals:
        raise AxisError(
            f"min_decimals {min_decimals} exceeds max_decimals {max_decimals}"
        )
    if max_decimals > 15:
        raise AxisError(f"max_decimals must be <= 15, got {max_decimals}")

    span = data_max - data_min
    flat = span <= abs(data_max) * _EPS

    if flat:  # R1
        if data_max == 0.0:
            lo, hi = -ZERO_SPAN, ZERO_SPAN
        else:
            half = abs(data_max) * FLAT_RATIO
            lo, hi = data_max - half, data_max + half
    else:  # R2
        pad = span * padding
        lo, hi = data_min - pad, data_max + pad

    min_step = _pow10(-max_decimals)
    step = nice_step((hi - lo) / (max_ticks - 1))

    axis_min = axis_max = 0.0
    count = 0
    for _ in range(128):  # R4, R5
        axis_min = math.floor(lo / step) * step
        axis_max = math.ceil(hi / step) * step
        while axis_min > data_min:
            axis_min -= step
        while axis_max < data_max:
            axis_max += step
        count = int(round((axis_max - axis_min) / step)) + 1
        if count <= max_ticks and step >= min_step * (1.0 - _EPS):
            break
        step = next_step(step)
    else:  # pragma: no cover - unreachable, the step at least doubles each pass
        raise AxisError("no tick solution found")

    while count < max_ticks:  # R6
        gap_low = data_min - axis_min
        gap_high = axis_max - data_max
        if abs(gap_low - gap_high) <= step / 2.0 * (1.0 + _EPS):
            break
        if gap_low < gap_high:
            axis_min -= step
        else:
            axis_max += step
        count += 1

    decimals = -_exponent(step)  # R7
    decimals = max(0, min(decimals, max_decimals))
    decimals = max(decimals, min_decimals)

    ticks = []
    labels = []
    for index in range(count):  # R8
        value = round(axis_min + index * step, decimals)
        if value == 0.0:
            value = 0.0  # collapse -0.0
        ticks.append(value)
        labels.append(f"{value:.{decimals}f}")

    return Axis(
        min=ticks[0],
        max=ticks[-1],
        step=step,
        decimals=decimals,
        ticks=tuple(ticks),
        labels=tuple(labels),
        flat=flat,
    )


def check_invariants(
    axis: Axis,
    data_min: float,
    data_max: float,
    max_ticks: int,
    min_decimals: int,
    max_decimals: int,
) -> Sequence[str]:
    """Return the post-conditions of SPEC.md that this axis violates."""
    failures = []
    step = axis.step
    tol = step * 1e-9

    for tick in axis.ticks:
        # Tolerance has to scale with the tick, not the step: at a magnitude of 1e6 with a
        # step of 1e-6 the ratio alone consumes twelve of a double's sixteen digits.
        slack = max(abs(tick), step) * 1e-11
        if abs(tick - round(tick / step) * step) > slack:
            failures.append(f"I1 tick {tick} is not a multiple of {step}")
            break
    if axis.min > data_min + tol or axis.max < data_max - tol:
        failures.append(f"I2 axis [{axis.min}, {axis.max}] misses [{data_min}, {data_max}]")
    mantissa = step / _pow10(_exponent(step))
    if not any(abs(mantissa - m) < 1e-9 for m in MANTISSAS):
        failures.append(f"I3 step {step} is not 1, 2 or 5 times a power of ten")
    if not 2 <= axis.count <= max_ticks:
        failures.append(f"I4 tick count {axis.count} outside [2, {max_ticks}]")
    if len(set(axis.labels)) != len(axis.labels):
        failures.append(f"I5 duplicate labels in {axis.labels}")
    if not min_decimals <= axis.decimals <= max_decimals:
        failures.append(f"I6 decimals {axis.decimals} outside [{min_decimals}, {max_decimals}]")
    gap_low = data_min - axis.min
    gap_high = axis.max - data_max
    if abs(gap_low - gap_high) > step / 2.0 * (1.0 + 1e-9) and axis.count != max_ticks:
        failures.append(
            f"I7 gaps {gap_low} and {gap_high} differ by more than half a step "
            f"with room to spare ({axis.count} of {max_ticks} ticks)"
        )
    for i in range(1, axis.count):
        delta = axis.ticks[i] - axis.ticks[i - 1]
        if delta <= 0 or abs(delta - step) > max(tol, _pow10(-axis.decimals) / 2):
            failures.append(f"I8 spacing {delta} at index {i} is not a clean {step}")
            break
    return failures

File: js/price-axis.mjs
/**
 * Deterministic y-axis ticks for price charts.
 *
 * Independent implementation of SPEC.md. No dependencies. It shares no code with the
 * Python reference; vectors.json is the only thing the two have in common.
 */

export const SPEC_VERSION = "1.0.0";

const MANTISSAS = [1, 2, 5];
const DEFAULT_MAX_TICKS = 8;
const DEFAULT_PADDING = 0.05;
const DEFAULT_MAX_DECIMALS = 6;
const FLAT_RATIO = 0.001;
const ZERO_SPAN = 1;

const EPS = 1e-12;
const LOG_NUDGE = 1e-9;

export class AxisError extends Error {
  constructor(message) {
    super(message);
    this.name = "AxisError";
  }
}

/**
 * 10**exp via the decimal literal. Math.pow is not guaranteed to return the correctly
 * rounded double for every exponent, so it can disagree with another language in the
 * last bit. The Python reference uses float(f"1e{exp}") for the same reason.
 */
function pow10(exp) {
  return Number(`1e${exp}`);
}

function exponentOf(value) {
  return Math.floor(Math.log10(value) + LOG_NUDGE);
}

/** Smallest step of the form m*10^n, m in {1,2,5}, that is >= raw. R3. */
export function niceStep(raw) {
  if (!Number.isFinite(raw) || raw <= 0) {
    throw new AxisError(`raw step must be finite and positive, got ${raw}`);
  }
  const exp = Math.floor(Math.log10(raw));
  for (const mantissa of MANTISSAS) {
    const candidate = mantissa * pow10(exp);
    if (candidate >= raw * (1 - EPS)) return candidate;
  }
  return pow10(exp + 1);
}

/** The next nice step up: 1 -> 2 -> 5 -> 10. R5. */
export function nextStep(step) {
  if (!Number.isFinite(step) || step <= 0) {
    throw new AxisError(`step must be finite and positive, got ${step}`);
  }
  const exp = exponentOf(step);
  const mantissa = step / pow10(exp);
  if (mantissa < 1.5) return 2 * pow10(exp);
  if (mantissa < 3.5) return 5 * pow10(exp);
  return pow10(exp + 1);
}

function checkNumber(name, value) {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new AxisError(`${name} must be a finite number, got ${value}`);
  }
  return value;
}

function checkCount(name, value, minimum) {
  if (!Number.isInteger(value) || value < minimum) {
    throw new AxisError(`${name} must be an integer >= ${minimum}, got ${value}`);
  }
  return value;
}

/**
 * Compute the axis for a price series. See SPEC.md.
 *
 * @returns {{min:number, max:number, step:number, decimals:number,
 *            ticks:number[], labels:string[], flat:boolean}}
 */
export function computeAxis(dataMin, dataMax, options = {}) {
  const {
    maxTicks = DEFAULT_MAX_TICKS,
    padding = DEFAULT_PADDING,
    minDecimals = 0,
    maxDecimals = DEFAULT_MAX_DECIMALS,
  } = options;

  checkNumber("dataMin", dataMin);
  checkNumber("dataMax", dataMax);
  checkNumber("padding", padding);

  if (dataMax < dataMin) {
    throw new AxisError(`dataMax ${dataMax} is below dataMin ${dataMin}`);
  }
  checkCount("maxTicks", maxTicks, 3);
  if (padding < 0) throw new AxisError(`padding must not be negative, got ${padding}`);
  checkCount("minDecimals", minDecimals, 0);
  checkCount("maxDecimals", maxDecimals, 0);
  if (minDecimals > maxDecimals) {
    throw new AxisError(`minDecimals ${minDecimals} exceeds maxDecimals ${maxDecimals}`);
  }
  if (maxDecimals > 15) {
    throw new AxisError(`maxDecimals must be <= 15, got ${maxDecimals}`);
  }

  const span = dataMax - dataMin;
  const flat = span <= Math.abs(dataMax) * EPS;

  let lo;
  let hi;
  if (flat) {
    // R1
    if (dataMax === 0) {
      lo = -ZERO_SPAN;
      hi = ZERO_SPAN;
    } else {
      const half = Math.abs(dataMax) * FLAT_RATIO;
      lo = dataMax - half;
      hi = dataMax + half;
    }
  } else {
    // R2
    const pad = span * padding;
    lo = dataMin - pad;
    hi = dataMax + pad;
  }

  const minStep = pow10(-maxDecimals);
  let step = niceStep((hi - lo) / (maxTicks - 1));

  let axisMin = 0;
  let axisMax = 0;
  let count = 0;
  let solved = false;
  for (let attempt = 0; attempt < 128; attempt += 1) {
    // R4, R5
    axisMin = Math.floor(lo / step) * step;
    axisMax = Math.ceil(hi / step) * step;
    while (axisMin > dataMin) axisMin -= step;
    while (axisMax < dataMax) axisMax += step;
    count = Math.round((axisMax - axisMin) / step) + 1;
    if (count <= maxTicks && step >= minStep * (1 - EPS)) {
      solved = true;
      break;
    }
    step = nextStep(step);
  }
  if (!solved) throw new AxisError("no tick solution found");

  while (count < maxTicks) {
    // R6
    const gapLow = dataMin - axisMin;
    const gapHigh = axisMax - dataMax;
    if (Math.abs(gapLow - gapHigh) <= (step / 2) * (1 + EPS)) break;
    if (gapLow < gapHigh) axisMin -= step;
    else axisMax += step;
    count += 1;
  }

  let decimals = -exponentOf(step); // R7
  decimals = Math.max(0, Math.min(decimals, maxDecimals));
  decimals = Math.max(decimals, minDecimals);

  const factor = pow10(decimals);
  const ticks = [];
  const labels = [];
  for (let index = 0; index < count; index += 1) {
    // R8
    let value = Math.round((axisMin + index * step) * factor) / factor;
    if (value === 0) value = 0; // collapse -0
    ticks.push(value);
    labels.push(value.toFixed(decimals));
  }

  return {
    min: ticks[0],
    max: ticks[ticks.length - 1],
    step,
    decimals,
    ticks,
    labels,
    flat,
  };
}

/** Return the post-conditions of SPEC.md that this axis violates. */
export function checkInvariants(axis, dataMin, dataMax, maxTicks, minDecimals, maxDecimals) {
  const failures = [];
  const { step } = axis;
  const tol = step * 1e-9;

  for (const tick of axis.ticks) {
    const slack = Math.max(Math.abs(tick), step) * 1e-11;
    if (Math.abs(tick - Math.round(tick / step) * step) > slack) {
      failures.push(`I1 tick ${tick} is not a multiple of ${step}`);
      break;
    }
  }
  if (axis.min > dataMin + tol || axis.max < dataMax - tol) {
    failures.push(`I2 axis [${axis.min}, ${axis.max}] misses [${dataMin}, ${dataMax}]`);
  }
  const mantissa = step / pow10(exponentOf(step));
  if (!MANTISSAS.some((m) => Math.abs(mantissa - m) < 1e-9)) {
    failures.push(`I3 step ${step} is not 1, 2 or 5 times a power of ten`);
  }
  if (!(axis.ticks.length >= 2 && axis.ticks.length <= maxTicks)) {
    failures.push(`I4 tick count ${axis.ticks.length} outside [2, ${maxTicks}]`);
  }
  if (new Set(axis.labels).size !== axis.labels.length) {
    failures.push(`I5 duplicate labels in ${axis.labels.join(", ")}`);
  }
  if (!(axis.decimals >= minDecimals && axis.decimals <= maxDecimals)) {
    failures.push(`I6 decimals ${axis.decimals} outside [${minDecimals}, ${maxDecimals}]`);
  }
  const gapLow = dataMin - axis.min;
  const gapHigh = axis.max - dataMax;
  if (
    Math.abs(gapLow - gapHigh) > (step / 2) * (1 + 1e-9) &&
    axis.ticks.length !== maxTicks
  ) {
    failures.push(
      `I7 gaps ${gapLow} and ${gapHigh} differ by more than half a step with room to spare`,
    );
  }
  for (let i = 1; i < axis.ticks.length; i += 1) {
    const delta = axis.ticks[i] - axis.ticks[i - 1];
    if (delta <= 0 || Math.abs(delta - step) > Math.max(tol, pow10(-axis.decimals) / 2)) {
      failures.push(`I8 spacing ${delta} at index ${i} is not a clean ${step}`);
      break;
    }
  }
  return failures;
}
